contrib/epee/include/syncobj.h

```diff
@@ -150,6 +154,127 @@ namespace epee
   };
 
 
+  struct reader_writer_lock {
+
+    reader_writer_lock() noexcept : 
+    write_mode(false),
+    read_mode(0) {};
+
+    bool start_read() noexcept {
+      if (have_write()) {
+        return false;
+      }
+      lock_reader();
+      return true;
+    }
+
+    void end_read() noexcept {
+      unlock_reader();
+      condition.notify_all();
+    }
+
+    bool start_write() noexcept {
+      if (have_write()) {
+        return false;
+      }
+      lock_and_change(boost::this_thread::get_id());
+      return true;
+    }
+
+    void end_write() noexcept {
+      unlock_writer();
+      condition.notify_all();
+    }
+
+    bool have_write() noexcept {
+      boost::mutex::scoped_lock lock(internal_mutex);
+      if (write_mode
+          && writer_id == boost::this_thread::get_id()) {
+        return true;
+      }
+      return false;
+    }
+
+    ~reader_writer_lock() = default;
+    reader_writer_lock(reader_writer_lock&) = delete;
+    reader_writer_lock operator=(reader_writer_lock&) = delete;
+
+  private:
+
+    void unlock_writer() noexcept {
+      CHECK_AND_ASSERT_MES2(write_mode && !read_mode, "Ending write transaction by " << boost::this_thread::get_id() << " while the lock is not in write mode");
+      boost::mutex::scoped_lock lock(internal_mutex);      
+      rw_mutex.unlock();
+      write_mode = false;
+      writer_id = boost::thread::id();
+    }
+
+    void unlock_reader() noexcept {
+      CHECK_AND_ASSERT_MES2(read_mode && !write_mode, "Ending read transaction by " << boost::this_thread::get_id() << " while the lock is not in read mode");
+      boost::mutex::scoped_lock lock(internal_mutex);      
+      rw_mutex.unlock_shared();
+      read_mode--;
+      if (rw_mutex.try_lock()) {
+        CHECK_AND_ASSERT_MES2(!read_mode, "Reader is not zero but goes to read_mode = 0 by " << boost::this_thread::get_id());
+        write_mode = false;
+        rw_mutex.unlock();
+      }
+      return; 
+    }
+
+    void lock_reader() noexcept {
+      CHECK_AND_ASSERT_MES2(read_mode < UINT32_MAX, "Maximum number of readers reached.");
```

Comment:
Again I am using `CHECK_AND_ASSERT_MES2` and I believe it does not throw. So even if assert fails, it only logs and moves ahead [1].

One thing to keep in mind is, that the check for the `UINT32_MAX` number of readers is auxiliary. 

For example, the Linux kernel uses this function to calculate the maximum number of threads allowed [2] and its value is `int` [3]. I cannot think of any scenario in which we can surpass `UINT32_MAX`.

1. https://github.com/monero-project/monero/blob/7b7958bbd9d76375c47dc418b4adabba0f0b1785/contrib/epee/include/misc_log_ex.h#L205C64-L205C73
2. https://github.com/torvalds/linux/blob/04b8076df2534f08bb4190f90a24e0f7f8930aca/kernel/fork.c#L1014
3. https://github.com/torvalds/linux/blob/04b8076df2534f08bb4190f90a24e0f7f8930aca/kernel/fork.c#L132

Reply:
Wait, `CHECK_AND_ASSERT_MES2` doesn't `return` immediately? Whereas `CHECK_AND_ASSERT_MES_NO_RET` does? Confusing.

And I still think you have to perform the check after acquiring the mutex.

> For example, the Linux kernel uses this function to calculate the maximum number of threads allowed [2] and its value is int [3]. I cannot think of any scenario in which we can surpass UINT32_MAX.

What's the purpose of the check then? Just to inform the user? The only other option is to spin loop block.

Reply:
> And I still think you have to perform the check after acquiring the mutex.

The reason is the mutex is blocking. and for the debugging, I wanted to see the checks before blocks. But this is one of the cases that I am open to either way. I will put it on my TODO list. In the next push will move the check inside the lock. 

> Just to inform the user? 
Yes. Exactly. Since debugging locks is hard and tricky I left it there if in **_an extremely_** rare scenario, in 15 years, if the user hits this problem, at least we can see it in the log files. I am open to removing that check too, since not directly related to core logic of the lock.

Reply:
Even if this check is auxiliary, technically `read_mode` either needs to be atomic, or the check needs to happen when `internal_mutex` is owned, if you want the loaded value of `read_mode` to be well-defined.

Reply:
Thanks @vtnerd  @jeffro256 . Fixed in new push.